zilliqa/src/consensus.rs

```diff
@@ -1056,11 +1109,61 @@ impl Consensus {
         // need those transactions again
         for tx in opaque_transactions {
             let account_nonce = self.state.get_account(tx.signer)?.nonce;
-            self.transaction_pool.insert_transaction(tx, account_nonce);
+            transaction_pool.insert_transaction(tx, account_nonce);
         }
         Ok(Some((proposal, broadcasted_transactions)))
     }
 
+    fn propose_new_block(&mut self) -> Result<Option<(Block, Vec<VerifiedTransaction>)>> {
+        let mut state = self.state.clone();
+        let mut tx_pool = self.transaction_pool.clone();
+        let mut votes = self.votes.clone();
+
+        let result = self.propose_new_block_at(&mut state, &mut tx_pool, &mut votes);
```

Comment:
I'm not super keen on cloning the pool and the votes on every proposal (cloning the state is cheap) - but the `&self` immutable borrow does make it difficult to just pass `&mut self.state, &mut self.transaction_pool, &mut self.votes` directly. Some of the usages of `&self` inside `propose_new_block()` can be eliminated, but it'd require 3-4 extra paremeters which gets ugly, so I'm not sure I have a better solution to suggest.

In a perfect world we would have a way to peek in the transaction pool immutably, similarly to how we can edit state and rollback painlessly, thus removing the need for `_at` in general (and mutating the votes is fine because it's just adding a default value if there isn't one). But I suppose this is a fine alternative for now, without having to redesign the transaction pool completely to allow that.